* info

** hg38 mm39 rheMac10 rn6
http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
http://hgdownload.soe.ucsc.edu/goldenPath/mm39/bigZips/mm39.fa.gz
http://hgdownload.soe.ucsc.edu/goldenPath/rheMac10/bigZips/rheMac10.fa.gz
http://hgdownload.soe.ucsc.edu/goldenPath/rn6/bigZips/rn6.fa.gz

** set path
export PATH=/hive/users/markd/gencode/projs/isoseq-xspecies/isoseq-xspecies/bin:${PATH}
** hub
https://hgwdev.gi.ucsc.edu/~markd/gencode/isoseq-xspecies/hub/hub.txt
http://genome.ucsc.edu/cgi-bin/hgGateway?db=hg38&hubUrl=https://hgwdev.gi.ucsc.edu/~markd/gencode/isoseq-xspecies/hub/hub.txt

* map annotations between all genomes:
2021-01-26
** download below files to build/annot-map/data/${db}/

ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_35/gencode.v35.annotation.gtf.gz
ftp://ftp.ensembl.org/pub/release-101/gtf/macaca_mulatta/Macaca_mulatta.Mmul_10.101.chr.gtf.gz
http://ftp.ebi.ac.uk/pub/databases/havana/gencode_pre/ensembl_ftp_files/ens_103_mouse_tmp/Mus_musculus.GRCm39.103.gtf.gz
ftp://ftp.ensembl.org/pub/release-101/gtf/rattus_norvegicus/Rattus_norvegicus.Rnor_6.0.101.gtf.gz        

** build mappings
cd build/annot-map
make -j 32

# filter was:
pslCDnaFilter -minCover=0.50 -bestOverlap -filterWeirdOverlapped

** link to save version
find chains data mapped |cpio -pduml versions/1.0/

* 2020-12-01 initial isoseq from Anoushka
cd  ../tmp/SeparateSpeciesAlignment_Minimap2_UCSCrefGenome/
human.bed
macaque.bed
mouse.bed
rat.bed

buildBigBed hg38 ../../hub/hg38/isoseq.minimap2.2020-12-02.bigBed human.bed 
buildBigBed rheMac10 ../../hub/rheMac10/isoseq.minimap2.2020-12-02.bigBed macaque.bed 
buildBigBed mm39 ../../hub/mm39/isoseq.minimap2.2020-12-02.bigBed mouse.bed
buildBigBed rn6  ../../hub/rn6/isoseq.minimap2.2020-12-02.bigBed rat.bed 

** 2021-01-26 remap annotations
# avoid problems with losing fragments due to synteny break with having
# a min cover filter. Change to:
#  pslCDnaFilter -localNearBest=0.001 -bestOverlap -filterWeirdOverlapped

cd build/annot-map
nice make -j 64 >&log&
find chains data mapped |cpio -pduml versions/1.1/
* 2021-02-23 exon conservation status on hg38 from Anoushka
cd hg38/exon_cons
conserved.bed  onlyHuman.bed  onlyPrimates.bed
for b in *.bed ; do bedToBigBed -type=bed9 $b /hive/data/genomes/hg38/chrom.sizes $(basename $b .bed).bigBed ; done
* 2021-03-21 rebuild without retained introns (V3.1)
** get official GRCm39
ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M26/gencode.vM26.annotation.gtf.gz
* 2021-05-04 repeat maskers as BED
in hub/download/repeat-master/

/hive/users/markd/nanopore/projs/t2t-chm13/T2T-CHM13-hub/bin/RM2Bed.py /hive/data/genomes/hg38/bed/repeatMasker/hg38.sorted.fa.out ../hub/download/repeat-master/hg38.repeatMasker.bed&
/hive/users/markd/nanopore/projs/t2t-chm13/T2T-CHM13-hub/bin/RM2Bed.py /hive/data/genomes/mm39/bed/repeatMasker/mm39.sorted.fa.out ../hub/download/repeat-master/mm39.repeatMasker.bed >&tmp/mm39.out&
/hive/users/markd/nanopore/projs/t2t-chm13/T2T-CHM13-hub/bin/RM2Bed.py /hive/data/genomes/rheMac10/bed/repeatMasker/rheMac10.sorted.fa.out ../hub/download/repeat-master/rheMac10.repeatMasker.bed >&tmp/rheMac10.out&
/hive/users/markd/nanopore/projs/t2t-chm13/T2T-CHM13-hub/bin/RM2Bed.py /hive/data/genomes/rn6/bed/repeatMasker/rn6.sorted.fa.out ../hub/download/repeat-master/rn6.repeatMasker.bed >&tmp/rn6.out&
# should have sorted, as not all of these are actually sorted.
* 2021-05-18 rebuild for bug
- bug fix for mapping multiple src exons to the same target exon
